c2rust-analyze/src/main.rs

```diff
@@ -482,6 +500,24 @@ fn run(tcx: TyCtxt) {
         }
     }
 
+    // FIX the fields of structs mentioned in extern blocks
+    for adt_did in &gacx.adt_metadata.struct_dids {
+        let ty = &tcx.type_of(adt_did);
+        if gacx.foreign_mentioned_tys.contains(ty) {
```

Comment:
The `Ty`s at this stage should still be generic, right?

Reply:
The `Ty` produced by `tcx.type_of(adt_did)` will be generic.  The `Ty`s in the `foreign_mentioned_tys` set won't be, since they come from actual uses of the struct.

Reply:
Ah, okay.  Then I agree that using `DefId`s seems smarter here.

Reply:
How would I get `DefId`s of function signature inputs and outputs, which are an iterator over types, to add those to the set?

Reply:
For each arg/return/field type, I think you'd want to walk over it (maybe with [`Ty::walk`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.walk) or `LTy::iter` or a custom recursive function) and do something with all the `TyKind::Adt` `DefId`s you see.  You could potentially also look for `PointerId`s and set their FIXED flags in the same traversal, which might be simpler depending on how exactly you implement the traversal of struct fields.

Reply:
Thanks, that looks very useful

Reply:
Walking in 795438ae